light-clients/common/src/lib.rs

```diff
@@ -65,25 +73,37 @@ where
 	H: hash_db::Hasher<Out = H256> + Debug + 'static,
 {
 	if root.as_bytes().len() != 32 {
-		return Err(anyhow!("invalid commitment root length: {}", root.as_bytes().len()))
+		return Err(anyhow!("invalid commitment root length: {}", root.as_bytes().len()));
 	}
 	let path: Path = path.into();
 	let path = path.to_string();
 	let mut key = prefix.as_bytes().to_vec();
 	key.extend(path.as_bytes());
-	let trie_proof: Vec<Vec<u8>> = codec::Decode::decode(&mut &*proof.as_bytes())
+	let ibc_proof: IbcProof = codec::Decode::decode(&mut &*proof.as_bytes())
 		.map_err(|err| anyhow!("Failed to decode proof nodes for path: {path}: {err:#?}"))?;
-	let proof = StorageProof::new(trie_proof);
 	let root = H256::from_slice(root.as_bytes());
 	let child_info = ChildInfo::new_default(prefix.as_bytes());
-	state_machine::read_child_proof_check::<H, _>(
-		root.into(),
-		proof,
-		child_info,
-		vec![(key, Some(value))],
-	)
-	.map_err(|err| anyhow!("Failed to verify proof for path: {path}, error: {err:#?}"))?;
-	Ok(())
+	let binding = child_info.prefixed_storage_key();
+	let trie_key = binding.as_slice();
+	// todo: get child root
+	let child_root_proof = &ibc_proof.child_trie_root_proof;
+	let child_proof = &ibc_proof.child_trie_proof;
+
+	let verify_root = verify_trie_proof::<LayoutV0<H>, _, _, _>(
+		&root.into(),
+		child_root_proof,
+		&[(trie_key, Some(&value))],
+	);
+	let verify_path = verify_trie_proof::<LayoutV0<H>, _, _, _>(
+		&root.into(),
+		child_proof,
+		&[(key, Some(&value))],
+	);
+
+	match (verify_root, verify_path) {
+		(Ok(_), Ok(_)) => Ok(()),
+		_ => Err(anyhow!("verification failed")),
```

Comment:
is there any value in discriminating which proof failed?

Reply:
definitely